### PR TITLE
feat(planner): drag-n-drop

### LINKED ISSUE #5001: Make todos sortable using drag and drop
<sub>View in Huly <a href="https://front.hc.engineering/guest/platform?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NWY3MDljMjRlMjUzNGIzYmU0MGJhMmMiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InBsYXRmb3JtIiwicHJvZHVjdElkIjoiIn0.7vyTjXVyNRhrZTexJJLIdRvHBg-XhkAlxaFevLrEkXA">UBERF-6064</a></sub>

### PR DESCRIPTION
**Pull request by euaaaio**
# Contribution checklist

### Screencast
https://github.com/hcengineering/platform/assets/5614115/72c9ffec-4cca-43c5-9aba-aa3b47c9e6d7

## Checklist

* [ ] - UI test added to added/changed functionality?
* [ ] - Screenshot is added to PR if applicable ?
* [x] - Does a local formatting is applied (rush format)
* [x] - Does a local svelte-check is applied (rush svelte-check)
* [ ] - Does a local UI tests are executed [UI Testing](../tests/readme.md)
* [ ] - Does the code work? Check whether function and logic are correct.
* [ ] - Does Changelog.md is updated with changes?
* [ ] - Does the translations are up to date?
* [ ] - Does it well tested?
* [x] - Tested for Chrome.
* [x] - Tested for Safari.
* [ ] - Go through the changed code looking for typos, TODOs, commented LOCs, debugging pieces of code, etc.
* [ ] - Rebase your branch onto master and upstream branch
* [ ] - Is there any redundant or duplicate code?
* [ ] - Are required links are linked to PR?
* [ ] - Does new code is well documented ?

## Related issues

Resolves #5001 

## Contributor requirements

* [x] - Sign-off is provided. [DCO](https://github.com/apps/dco)
* [x] - GPG Signature is provided. [GPG](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits)

<sub>View in Huly <a href="https://front.hc.engineering/guest/platform?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NWZjMWQ0YzkyYjg2MmVjNDNhZDA4ZjgiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InBsYXRmb3JtIiwicHJvZHVjdElkIjoiIn0.PAQKV5ZSYboYErCcOnzE5CH-uTCUE2TIrL3r69bybPM">UBERF-6125</a></sub>

### CODE REVIEW

File: models/time/src/migration.ts [hunk 1]
```diff
@@ -57,7 +58,8 @@ export async function migrateWorkSlots (client: TxOperations): Promise<void> {
             workslots: 0,
             priority: ToDoPriority.NoPriority,
             user: account.person,
-            visibility: 'public'
+            visibility: 'public',
+            rank: '' // TODO
```
Comment (haiodo) [inline models/time/src/migration.ts]: Migration should calculate them in some proper order, not just ''

File: models/time/src/migration.ts [hunk 2]
```diff
@@ -105,6 +107,44 @@ async function migrateTodosSpace (client: TxOperations): Promise<void> {
...
+    time.class.ToDo,
+    {
+      rank: { $exists: false },
+      doneOn: { $ne: null }
+    },
+    {
+      sort: { doneOn: SortingOrder.Ascending }
+    }
+  )
+  let undoneTodoRank = makeRank(undefined, undefined)
```
Comment (haiodo) [inline models/time/src/migration.ts:141]: I suppose we need to organize user todos as they for now per individual employee in system. And make a proper ranks, not just random ones.

File: plugins/document-resources/src/components/node-view/ToDoItemNodeView.svelte [hunk 3]
```diff
@@ -75,6 +76,23 @@
       await ops.remove(todo)
     }
 
+    const doneOn = node.attrs.checked === true ? Date.now() : null
+
+    const latestTodoItem = (
+      await ops.findAll(
```
Comment (haiodo) [inline plugins/document-resources/src/components/node-view/ToDoItemNodeView.svelte]: a findOne could be used.

File: plugins/time-resources/src/components/CreateToDo.svelte [hunk 4]
```diff
@@ -17,13 +18,27 @@
     description = description?.trim() ?? ''
     const client = getClient()
     const acc = getCurrentAccount() as PersonAccount
+    const latestTodo = (
+      await client.findAll(
```
Comment (haiodo) [inline plugins/time-resources/src/components/CreateToDo.svelte]: a findOne could be used

File: plugins/time-resources/src/components/CreateToDoPopup.svelte [hunk 5]
```diff
@@ -54,6 +55,19 @@
 
   async function saveToDo (): Promise<void> {
     loading = true
+    const latestTodo = (
+      await client.findAll(
```
Comment (haiodo) [inline plugins/time-resources/src/components/CreateToDoPopup.svelte]: fineOne

File: plugins/time/src/index.ts [hunk 6]
```diff
@@ -36,7 +37,14 @@ export interface WorkSlot extends Event {
 /**
  * @public
  */
-export interface ToDo extends AttachedDoc {
+export interface DocWithRank extends Doc {
```
Comment (haiodo) [inline plugins/time/src/index.ts]: Why we need this change?

File: models/time/src/migration.ts (hunk 2 above)
Comment (haiodo) [reply models/time/src/migration.ts:141]: ok, not a random.

File: models/time/src/migration.ts (hunk 1 above)
Comment (euaaaio) [reply models/time/src/migration.ts]: Done

File: plugins/document-resources/src/components/node-view/ToDoItemNodeView.svelte (hunk 3 above)
Comment (euaaaio) [reply plugins/document-resources/src/components/node-view/ToDoItemNodeView.svelte]: Done

File: plugins/time-resources/src/components/CreateToDo.svelte (hunk 4 above)
Comment (euaaaio) [reply plugins/time-resources/src/components/CreateToDo.svelte]: Done

File: plugins/time-resources/src/components/CreateToDoPopup.svelte (hunk 5 above)
Comment (euaaaio) [reply plugins/time-resources/src/components/CreateToDoPopup.svelte]: Done

File: plugins/time/src/index.ts (hunk 6 above)
Comment (euaaaio) [reply plugins/time/src/index.ts]: Forgot to remove it

**Review by haiodo — APPROVED** [review]

**Pull request merged**